src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/SNI/SslOverTdsStream.NetCoreApp.cs

```diff
@@ -0,0 +1,289 @@
+                    int packetBytesRead;
+                    {
+                        ValueTask<int> packetReadValueTask = _stream.ReadAsync(buffer.Slice(0, Math.Min(buffer.Length, _packetBytes)), cancellationToken);
+                        if (packetReadValueTask.IsCompletedSuccessfully)
+                        {
+                            packetBytesRead = packetReadValueTask.Result;
+                        }
+                        else
+                        {
+                            packetBytesRead = await packetReadValueTask.AsTask().ConfigureAwait(false);
+                        }
+                    }
+                    _packetBytes -= packetBytesRead;
+                    return packetBytesRead;
+                }
+            }
+            else
+            {
+                int read;
+                {
+                    ValueTask<int> readValueTask = _stream.ReadAsync(buffer, cancellationToken);
+                    if (readValueTask.IsCompletedSuccessfully)
+                    {
+                        read = readValueTask.Result;
+                    }
+                    else
+                    {
+                        read = await readValueTask.AsTask().ConfigureAwait(false);
+                    }
+                }
+                return read;
+            }
+        }
+
+        public override void Write(ReadOnlySpan<byte> buffer)
+        {
+            // During the SSL negotiation phase, SSL is tunnelled over TDS packet type 0x12. After
+            // negotiation, the underlying socket only sees SSL frames.
+            if (_encapsulate)
+            {
+                ReadOnlySpan<byte> remaining = buffer;
+                byte[] packetBuffer = null;
+                while (remaining.Length > 0)
+                {
+                    int dataLength = Math.Min(PACKET_SIZE_WITHOUT_HEADER, remaining.Length);
+                    int packetLength = TdsEnums.HEADER_LEN + dataLength;
+
+                    if (packetBuffer == null)
+                    {
+                        packetBuffer = ArrayPool<byte>.Shared.Rent(packetLength);
+                    }
+                    else if (packetBuffer.Length < packetLength)
+                    {
+                        ArrayPool<byte>.Shared.Return(packetBuffer, clearArray: true);
+                        packetBuffer = ArrayPool<byte>.Shared.Rent(packetLength);
+                    }
+
+                    SetupPreLoginPacketHeader(packetBuffer, dataLength, remaining.Length - dataLength);
+
+                    Span<byte> data = packetBuffer.AsSpan(TdsEnums.HEADER_LEN, dataLength);
+                    remaining.Slice(0, dataLength).CopyTo(data);
+
+                    _stream.Write(packetBuffer.AsSpan(0, packetLength));
+                    _stream.Flush();
+
+                    remaining = remaining.Slice(dataLength);
+                }
+                if (packetBuffer != null)
```

Comment:
I'd place this in a finally clause, to also return the buffer in case of exceptions. IIRC this isn't an actual memory leak (unreturned buffers do get reclaimed by the GC), but this would be more correct (and allow buffers to get reused even in case of an exception).

Reply:
This is just me being cautious. I tend to be quite security concious which leads me to handle exceptions as catastrphic. In this case if an error has occured i don't want to try and rescue the buffer back into the pool because i don't know if the error occured because someone was messing with the buffer. If we hit an exception just throw everything away as dirty. 

If there's a particular preference for a try-finally block just say and i'll change it.

Reply:
I'm not sure why an exception would be more related to a security concern than a non-exceptional flow - in both cases we've read some SQL Server TDS data into the buffer, either it's OK to return it as-is to the pool, or it isn't... Note that ArrayPool.Return [has an option](https://docs.microsoft.com/en-us/dotnet/api/system.buffers.arraypool-1.return?view=netcore-3.1#System_Buffers_ArrayPool_1_Return__0___System_Boolean_) for clearing the buffer (off by default) if that's something you want to do. But I'd think more broadly about what kind of security policy/guarantees you want to provide.